### PR TITLE
Signature algorithm refactor, part 3.

### PR DESCRIPTION
**Pull request by snhenson**
##### Checklist

- [x] tests are added or updated

##### Description of change

This PR adds support for TLS 1.3 style signature scheme based certificate selection. It saves the index of the TLS 1.3 certificate we use for later retrieval. Although this technique is currently only used with TLS 1.3 it could be applied to earlier versions: potentially tidying up the certificate selection code considerably... but that's for a later PR.

### CODE REVIEW

File: ssl/t1_lib.c [hunk 1]
```diff
@@ -2270,3 +2270,59 @@ int ssl_security_cert_chain(SSL *s, STACK_OF(X509) *sk, X509 *x, int vfy)
...
+ * Choose an appropriate signature algorithm based on available certificates
+ * Set current certificate and digest to match chosen algorithm.
+ */
+
+int tls_choose_sigalg(SSL *s)
+{
+    if (SSL_IS_TLS13(s)) {
+        size_t i;
+        int curve = -1;
+        /* Look for a certificate matching shared sigaglgs */
```
Comment (richsalz) [inline ssl/t1_lib.c]: blank line after decl

File: ssl/t1_lib.c [hunk 2]
```diff
@@ -2270,3 +2270,59 @@ int ssl_security_cert_chain(SSL *s, STACK_OF(X509) *sk, X509 *x, int vfy)
...
+            s->s3->tmp.sigalg = lu;
+            s->s3->tmp.cert_idx = idx;
+            s->s3->tmp.md[idx] = md;
+            s->cert->key = s->cert->pkeys + idx;
+            return 1;
+        }
+        return 0;
+    }
+    /*
+     * FIXME: could handle previous TLS versions in an appropriate way
```
Comment (richsalz) [inline ssl/t1_lib.c]: if you think this FIXME will be handled soon, fine.  If not, maybe invert the opening if test, duplicate the "return 1" lines and move the whole function body back one tabstop?

File: ssl/t1_lib.c [hunk 3]
```diff
@@ -2270,3 +2270,59 @@ int ssl_security_cert_chain(SSL *s, STACK_OF(X509) *sk, X509 *x, int vfy)
...
+            CERT_PKEY *c;
+
+            /* Skip RSA if not PSS */
+            if (lu->sig == EVP_PKEY_RSA)
+                continue;
+            md = ssl_md(lu->hash_idx);
+            if (md == NULL)
+                continue;
+            idx = lu->sig_idx;
+            c = s->cert->pkeys + idx;
```
Comment (richsalz) [inline ssl/t1_lib.c]: I prefer ```&s->cert->pkeys[idx]``` kind of thing, but not enough to push on the issue.

File: ssl/t1_lib.c [hunk 4]
```diff
@@ -2270,3 +2270,59 @@ int ssl_security_cert_chain(SSL *s, STACK_OF(X509) *sk, X509 *x, int vfy)
...
+                if (idx != SSL_PKEY_RSA_SIGN)
+                    continue;
+                idx = SSL_PKEY_RSA_ENC;
+                c = s->cert->pkeys + idx;
+                if (c->x509 == NULL || c->privatekey == NULL)
+                    continue;
+            }
+            if (lu->sig == EVP_PKEY_EC) {
+                if (curve == -1) {
+                    EC_KEY *ec = EVP_PKEY_get0_EC_KEY(c->privatekey);
```
Comment (richsalz) [inline ssl/t1_lib.c]: blank line after decl.  or merge into a single statement.

**Comment by vdukhovni** [comment]
Question: Does this yet deal with situations in which the client signature algorithms don't overlap with the server (only or any of) certificate? While we must to try choose a compatible chain when we can, we need to choose some default chain (typically the only one we have) when we can't. Does this or prior work address this issue?

File: ssl/t1_lib.c (hunk 2 above)
Comment (snhenson) [inline ssl/t1_lib.c]: Yes, that's the next thing I'm looking into. When looking at this I noted that the handling of certificate and signature selection is a mess and could be cleaned up by handling other versions of TLS in this function.

**Comment by snhenson** [comment]
This isn't currently addressed. If the client signature algorithms doesn't match any EE certificate: the handshake aborts with a fatal error. 

For TLS 1.3 we *could* change that to use whatever chain is available. Though we'd have to work out what logic to apply for multiple chains and which signature algorithm to use.

For versions prior to 1.3 it's trickier because the certificate type is tied to the ciphersuite.

File: ssl/t1_lib.c (hunk 4 above)
Comment (snhenson) [inline ssl/t1_lib.c]: Done

File: ssl/t1_lib.c (hunk 1 above)
Comment (snhenson) [inline ssl/t1_lib.c]: Done.

File: ssl/t1_lib.c (hunk 3 above)
Comment (snhenson) [inline ssl/t1_lib.c]: I don't care much either way so I changed it.

**Comment by richsalz** [comment]
For TLS 1.3, the abort is reasonable.  Yes you can send back whatever you have, but really, there are only two types of certs in TLS 1.3 :)

**Comment by vdukhovni** [comment]
Rich, the requirements for TLS 1.3 are clearly specified in https://tools.ietf.org/html/draft-ietf-tls-tls13-18#section-4.4.1.2 (top of page 56 in draft18):
```
   All certificates provided by the server MUST be signed by a signature
   algorithm that appears in the "signature_algorithms" extension
   provided by the client, if they are able to provide such a chain (see
   Section 4.2.3).  Certificates that are self-signed or certificates
   that are expected to be trust anchors are not validated as part of
   the chain and therefore MAY be signed with any algorithm.

   If the server cannot produce a certificate chain that is signed only
   via the indicated supported algorithms, then it SHOULD continue the
   handshake by sending the client a certificate chain of its choice
   that may include algorithms that are not known to be supported by the
   client.  This fallback chain MAY use the deprecated SHA-1 hash
   algorithm only if the "signature_algorithms" extension provided by
   the client permits it.  If the client cannot construct an acceptable
   chain using the provided certificates and decides to abort the
   handshake, then it MUST abort the handshake with an
   "unsupported_certificate" alert.

   If the server has multiple certificates, it chooses one of them based
   on the above-mentioned criteria (in addition to other criteria, such
   as transport layer endpoint, local configuration and preferences).
```
This is not restricted to TLS 1.3, similar text should have been in earlier TLS specifications.

**Comment by snhenson** [comment]
The current TLS 1.3 code decides which signature algorithm to use based on the available EE key and the preferences. It doesn't curently select using signature algorithms in the whole chain: so it would just send whatever chain was configured.

What that quoted text above dosn't cover is the case where the EE key doesn't match any supported signature algorithm. There the signature would be on TLS messages and not on certificates. That's where it currently aborts.

**Comment by vdukhovni** [comment]
> @snhenson What that quoted text above doesn't cover is the case where the EE key doesn't match any supported signature algorithm. There the signature would be on TLS messages and not on certificates. That's where it currently aborts.

No objection to aborting when the client's supported signature algorithms do not include any of the available EE public key types.  The intent of the text in the TLS 1.3 draft is to avoid needless aborts when potentially unsupported signatures are used in chain signatures. While we're obligated (when multiple chains are available) to try to select a chain that is compatible, when no chain is fully compatible, we must continue anyway and choose some default chain.  Simplest is to ignore the signature algorithm constraints and just choose the one that goes with the most preferred algorithms.

**Comment by richsalz** [comment]
Well, not to get into the "how many angels" aspect of things, but...  fallback is a SHOULD not MUST.  And in all these years while you say it "should have said that" nobody filed an errata or anything.

Whatever.  This is not a hill I'm going to die on, just pointing out that the current behavior if legit, even if some think it sub-optimal.

**Comment by vdukhovni** [comment]
We're not talking about fallback.  We're talking about avoiding foolishly refusing to complete a TLS handshake because the X.509 chain (which should be largely out of scope of the TLS spec) may or may not meet the client's TLS algorithm profile.  Many clients are doing opportunistic TLS and don't give a damn about the X.509 chain, but do want decent crypto. So it is really bad to refuse to do TLS, on an assumption of what the client might or might accept for X.509 auth.

**Comment by snhenson** [comment]
In case it isn't clear OpenSSL will never refuse to send a chain based on the client's preferences provided the EE key is acceptable so it does pretty much what @vdukhovni is suggesting already.

File: ssl/s3_lib.c [hunk 5]
```diff
@@ -838,8 +838,7 @@ static SSL_CIPHER ssl3_ciphers[] = {
      1,
      TLS1_3_TXT_AES_128_GCM_SHA256,
      TLS1_3_CK_AES_128_GCM_SHA256,
-     SSL_kRSA,
-     SSL_aRSA,
+     0, 0,
```
Comment (mattcaswell) [inline ssl/s3_lib.c]: I'm wondering if there is anything preventing us from adding the remaining TLSv1.3 MTI ciphersuites at this point.

File: test/ssl-tests/20-cert-select.conf.in [hunk 6]
```diff
@@ -104,3 +115,88 @@ our @tests = (
...
+        test   => {
+            "ExpectedServerCertType" => "RSA",
+            "ExpectedServerSignHash" => "SHA384",
+            "ExpectedServerSignType" => "RSA-PSS",
+            "ExpectedResult" => "Success"
+        },
+    },
+    {
+        name => "TLS 1.3 ECDSA Signature Algorithm Selection, no ECDSA certificate",
+        server => { },
```
Comment (mattcaswell) [inline test/ssl-tests/20-cert-select.conf.in]: Should this have the TLSv1.3 MinProtocol/MaxProtocol settings?

Comment (snhenson) [inline test/ssl-tests/20-cert-select.conf.in]: Yes, missed another couple for TLSv1.2 too. New commit pushed with a fix.

File: ssl/s3_lib.c (hunk 5 above)
Comment (snhenson) [inline ssl/s3_lib.c]: Can't think of any reason. I can do that in a separate PR or you can if you wish.

File: ssl/s3_lib.c [hunk 7]
```diff
@@ -3664,12 +3667,13 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
          * if we are considering an ECC cipher suite that uses an ephemeral
          * EC key check it
          */
```
Comment (mattcaswell) [inline ssl/s3_lib.c]: This comment and the fprintf a few lines above need indentation to be fixed

File: ssl/t1_lib.c [hunk 8]
```diff
@@ -2270,3 +2270,61 @@ int ssl_security_cert_chain(SSL *s, STACK_OF(X509) *sk, X509 *x, int vfy)
     }
     return 1;
 }
+
+/*
+ * Choose an appropriate signature algorithm based on available certificates
+ * Set current certificate and digest to match chosen algorithm.
+ */
+
```
Comment (mattcaswell) [inline ssl/t1_lib.c]: Lose this blank line

File: ssl/s3_lib.c (hunk 5 above)
Comment (mattcaswell) [inline ssl/s3_lib.c]: Happy to let you take that one :-)

File: test/ssl-tests/20-cert-select.conf.in [hunk 9]
```diff
@@ -104,3 +119,91 @@ our @tests = (
         },
     }
 );
+
+
+my $server_tls_1_3 = {
+    "ECDSA.Certificate" => "\${ENV::TEST_CERTS_DIR}${dir_sep}server-ecdsa-cert.pem",
+    "ECDSA.PrivateKey" => "\${ENV::TEST_CERTS_DIR}${dir_sep}server-ecdsa-key.pem",
+    "MinProtocol" => "TLSv1.3",
+    "MaxProtocol" => "TLSv1.3"
```
Comment (mattcaswell) [inline test/ssl-tests/20-cert-select.conf.in]: I'm wondering whether this should just be "MinProtocol", i.e. no "Maxprotocol". Wouldn't make any difference until TLSv1.4 though, so not a big deal.

**Comment by snhenson** [comment]
@richsalz you requested changes are you happy with this now?

**Comment by richsalz** [comment]
+1 / happy happy :)

**Pull request closed without merging**